src/graph/SetExecutor.cpp

```diff
@@ -0,0 +1,384 @@
+            if (!stat.ok()) {
+                return stat;
+            }
+       }
+       row.set_columns(std::move(cols));
+    }
+
+    return Status::OK();
+}
+
+
+std::vector<cpp2::RowValue> SetExecutor::doDistinct(
+        std::vector<cpp2::RowValue> &leftRows,
+        std::vector<cpp2::RowValue> &rightRows) const {
+    std::vector<cpp2::RowValue> rows;
+    for (auto &lr : leftRows) {
+        auto iter = rows.begin();
+        while (iter != rows.end()) {
+            if (lr == *iter) {
+                break;
+            }
+            ++iter;
+        }
+        if (iter == rows.end()) {
+            rows.emplace_back(std::move(lr));
+        }
+    }
+
+    for (auto &rr : rightRows) {
+        auto iter = rows.begin();
+        while (iter != rows.end()) {
+            if (rr == *iter) {
+                break;
+            }
+            ++iter;
+        }
+        if (iter == rows.end()) {
+            rows.emplace_back(std::move(rr));
+        }
+    }
+
+    return rows;
+}
+
+void SetExecutor::doIntersect() {
+    VLOG(3) << "Do InterSect.";
+    if (leftResult_ == nullptr || rightResult_ == nullptr) {
+        VLOG(3) << "No intersect.";
+        onEmptyInputs();
+        return;
+    }
+
+    Status status = checkSchema();
+    if (!status.ok()) {
+        DCHECK(onError_);
+        onError_(status);
+        return;
+    }
+
+    auto leftRows = leftResult_->getRows();
+    auto rightRows = rightResult_->getRows();
+    if (!castingMap_.empty()) {
+        Status stat = doCasting(rightRows);
+        if (!stat.ok()) {
+            DCHECK(onError_);
+            onError_(status);
+            return;
+        }
+    }
+
+    std::vector<cpp2::RowValue> rows;
+    for (auto &lr : leftRows) {
+        for (auto &rr : rightRows) {
+            if (rr == lr) {
+                rows.emplace_back(std::move(rr));
+                break;
+            }
+        }
+    }
+
+    finishExecution(std::move(rows));
+    return;
+}
+
+void SetExecutor::doMinus() {
+    VLOG(3) << "Do Minus.";
+    if (leftResult_ == nullptr) {
+        VLOG(3) << "Minus has only right result.";
+        onEmptyInputs();
+        return;
+    }
+
+    if (rightResult_ == nullptr) {
+        VLOG(3) << "Minus has left result.";
+        finishExecution(std::move(leftResult_));
+        return;
+    }
+
+    Status status = checkSchema();
+    if (!status.ok()) {
+        DCHECK(onError_);
+        onError_(status);
+        return;
+    }
+
+    auto leftRows = leftResult_->getRows();
+    auto rightRows = rightResult_->getRows();
+    if (!castingMap_.empty()) {
+        Status stat = doCasting(rightRows);
+        if (!stat.ok()) {
+            DCHECK(onError_);
+            onError_(status);
+            return;
+        }
+    }
+
+    for (auto &rr : rightRows) {
+        for (auto iter = leftRows.begin(); iter < leftRows.end();) {
+            if (rr == *iter) {
+                iter = leftRows.erase(iter);
```

Comment:
this class do a lot of op that insert or delete element from vector, but I think this is not a good ideal, it will take a lot memmove or memcopy when there are a lot of elements.

Reply:
That's right. But we can only try to reduce memmove or memcopy, due to earlier design in #667 . If you have a better idea to reduce these situation, or some optimization, please tell me!

Reply:
I think use list is the simple solution, and we need define a function return list in InterimResult. and more, I can not find list in our code anywhere except raft, why we just use vector?

Reply:
That would create conversion problems. list, string, vector...

Reply:
I made some changes here, please review the changes, thank you!